extensions/wikia/Email/EmailController.class.php

```diff
@@ -0,0 +1,292 @@
+<?php
+
+namespace Email;
+
+abstract class EmailController extends \WikiaController {
+	const DEFAULT_TEMPLATE_ENGINE = \WikiaResponse::TEMPLATE_ENGINE_MUSTACHE;
+
+	/** @var \User The user associated with the current request */
+	protected $currentUser;
+
+	/** @var \User The user this email will be sent to */
+	protected $targetUser;
+
+	protected $hasErrorResponse = false;
+
+	/** @var bool Whether or not to actually send an email */
+	protected $test;
+
+	/**
+	 * Since the children of this class are located in the 'Controller' directory, the default
+	 * location for the template directory would be 'Controller/templates'.  Redefine it to be
+	 * at the root of this extension instead.
+	 *
+	 * @return string
+	 */
+	public static function getTemplateDir() {
+		return dirname( __FILE__ ) . '/templates';
+	}
+
+	public function init() {
+		try {
+			$this->assertCanAccessController();
+
+			$this->currentUser = $this->findUserFromRequest( 'currentUser', $this->wg->User );
+			$this->targetUser = $this->findUserFromRequest( 'targetUser', $this->wg->User );
+			$this->test = $this->getRequest()->getVal( 'test', false );
+
+			$this->initEmail();
+		} catch ( ControllerException $e ) {
+			return $this->setErrorResponse( $e );
+		}
+	}
+
+	/**
+	 * Make sure to only allow authorized use of this extension.
+	 *
+	 * @throws \Email\Fatal
+	 */
+	public function assertCanAccessController() {
+		if ( $this->wg->User->isStaff() ) {
+			return;
+		}
+
+		if ( $this->request->isInternal() ) {
+			return;
+		}
+
+		throw new Fatal( wfMessage( 'emailext-error-restricted-controller' )->escaped() );
+	}
+
+	/**
+	 * Allow child classes to set some initial values.  Added here so its always defined
+	 * but is kept blank so does not need to be called via parent::initEmail()
+	 */
+	public function initEmail() {
```

Comment:
If it was an abstract function it would *have* to be defined in the child class, even if the child class didn't need to add any thing.